<a href="https://colab.research.google.com/github/zwen000/CS448_Assignment_4/blob/main/a3_sample_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

\(\^Be sure to update this button to point to your notebook instead of the sample notebook\)

In [207]:
# Imports and pip installations (if needed)
from sklearn.datasets import load_iris
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score,ShuffleSplit,cross_val_predict
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Part 1: Load the dataset

In [208]:
# Load the dataset (load remotely, not locally)
iris = load_iris()
# Output the first 15 rows of the data

# Display a summary of the table information (number of datapoints, etc.)
#df = pd.DataFrame(data= np.c_[iris['data'], iris['target']],columns= iris['feature_names'] + ['target'])
df=pd.DataFrame(np.c_[iris['data'], iris['target']])

df.columns=iris['feature_names']+ ['target']
df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
4      0.0  
5      0.0  
6      0.0  
7      0.0  
8      0.0  
9      0.0  
10     0.0  
11     0.0  
12     0.0  
13     0.0  
14     0.0

In [209]:
df.info()
df.describe()
df.size

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
sepal length (cm)    150 non-null float64
sepal width (cm)     150 non-null float64
petal length (cm)    150 non-null float64
petal width (cm)     150 non-null float64
target               150 non-null float64
dtypes: float64(5)
memory usage: 5.9 KB


750

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

# Part 2: Split the dataset into train and test

In [210]:
# Take the dataset and split it into our features (X) and label (y)
X = df[["sepal length (cm)", "sepal width (cm)", "petal length (cm)", "petal width (cm)"]]
y = df[["target"]]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state = 67)
X_train, X_test, y_train, y_test


(     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
 71                 6.1               2.8                4.0               1.3
 103                6.3               2.9                5.6               1.8
 61                 5.9               3.0                4.2               1.5
 17                 5.1               3.5                1.4               0.3
 68                 6.2               2.2                4.5               1.5
 3                  4.6               3.1                1.5               0.2
 72                 6.3               2.5                4.9               1.5
 79                 5.7               2.6                3.5               1.0
 48                 5.3               3.7                1.5               0.2
 9                  4.9               3.1                1.5               0.1
 108                6.7               2.5                5.8               1.8
 50                 7.0               3.2           

In [211]:
print(X_test.iloc[0])
sample = np.array([[ 5.0,3.5,1.6,0.6]])

sepal length (cm)    5.0
sepal width (cm)     3.5
petal length (cm)    1.6
petal width (cm)     0.6
Name: 43, dtype: float64


# Part 3: Logistic Regression

In [212]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression()
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
#sample = np.array([[ 5.1,3.8,1.5,0.3]])
y_prediction = model.predict(sample)
print(f"Sample datapoint X: {sample}, Prediction of  y: {y_prediction}")

probabilities = model.predict_proba(sample)
for i in range(len(probabilities[0])):
    print(f"The probability of sample data point in class {i} is {probabilities[0][i]}")


Sample datapoint X: [[5.  3.5 1.6 0.6]], Prediction of  y: [0.]
The probability of sample data point in class 0 is 0.9095881304179716
The probability of sample data point in class 1 is 0.09033855866140514
The probability of sample data point in class 2 is 7.331092062328637e-05


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


In [213]:
# iii. Report on the score for Logistic regression model, what does the score measure?
score = model.score(X_test, y_test)
print("Score of the model: ", score)

Score of the model:  0.9333333333333333


The score is the coefficient of prediction, it represents the model's capability in predicting the correct label. The best possible score is 1.0, a model with a score closer to 1.0 has high chance to predict correct label.
Our model has score 0.9333333333333333, very close to 1.0, that is suggesting that our model is fitted pretty well to the existing data, our predicted results will show relatively high accuracy.

In [214]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
intercept = model.intercept_
coef = model.coef_

print ("Intercept: ", intercept)
print ("Coefficents", coef)

Intercept:  [ 0.25502736  0.99848572 -1.05413089]
Coefficents [[ 0.41506347  1.40920057 -2.201387   -1.0143722 ]
 [ 0.36244035 -1.46622346  0.49522433 -1.2056787 ]
 [-1.68272696 -1.54678763  2.4891753   2.38360626]]


# Part 4: Support Vector Machine

In [215]:
# i. Use sklearn to train a Support Vector Classifier on the training set
model = svm.SVC(probability=True)
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
#sample = np.array([[ 5.1,3.8,1.5,0.3]])
y_prediction = model.predict(sample)
print(f"Sample datapoint X: {sample}, Prediction of  y: {y_prediction}")

probabilities = model.predict_proba(sample)
for i in range(len(probabilities[0])):
    print(f"The probability of sample data point in class {i} is {probabilities[0][i]}")

Sample datapoint X: [[5.  3.5 1.6 0.6]], Prediction of  y: [0.]
The probability of sample data point in class 0 is 0.9570514837619674
The probability of sample data point in class 1 is 0.026279548987724756
The probability of sample data point in class 2 is 0.01666896725030796


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


In [216]:
# iii. Report on the score for the SVM, what does the score measure?
score = model.score(X_test, y_test)
print("Score of the model: ", score)


Score of the model:  0.9333333333333333


The score is the coefficient of prediction, it represents the model's capability in predicting the correct label. The best possible score is 1.0, a model with a score closer to 1.0 has high chance to predict correct label.
Our model has score 0.9333333333333333, very close to 1.0, that is suggesting that our model is fitted pretty well to the existing data, our predicted results will show relatively high accuracy.

# Part 5: Neural Network

In [217]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
model = MLPClassifier()
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
#sample = np.array([[ 5.1,3.8,1.5,0.3]])
y_prediction = model.predict(sample)
print(f"Sample datapoint X: {sample}, Prediction of  y: {y_prediction}")

probabilities = model.predict_proba(sample)
for i in range(len(probabilities[0])):
    print(f"The probability of sample data point in class {i} is {probabilities[0][i]}")



Sample datapoint X: [[5.  3.5 1.6 0.6]], Prediction of  y: [0.]
The probability of sample data point in class 0 is 0.9270237582072944
The probability of sample data point in class 1 is 0.07218584657854854
The probability of sample data point in class 2 is 0.0007903952141568756


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:921: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


In [218]:
# iii. Report on the score for the Neural Network, what does the score measure?
score = model.score(X_test, y_test)
print("Score of the model: ", score)


Score of the model:  0.9333333333333333


The score is the coefficient of prediction, it represents the model's capability in predicting the correct label. The best possible score is 1.0, a model with a score closer to 1.0 has high chance to predict correct label.
Our model has score 0.9333333333333333, very close to 1.0, that is suggesting that our model is fitted pretty well to the existing data, our predicted results will show relatively high accuracy.

In [219]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)
model = MLPClassifier(hidden_layer_sizes=100)
model.fit(X_train, y_train)
score = model.score(X_test, y_test)
print("Score of the model: ", score)


Score of the model:  0.9333333333333333


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:921: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


# Part 6: K-Nearest Neighbors

In [220]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
model = KNeighborsClassifier()
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample = np.array([[ 5.1,3.8,1.5,0.3]])
y_prediction = model.predict(sample)
print(f"Sample datapoint X: {sample}, Prediction of  y: {y_prediction}")

probabilities = model.predict_proba(sample)
for i in range(len(probabilities[0])):
    print(f"The probability of sample data point in class {i} is {probabilities[0][i]}")



Sample datapoint X: [[5.1 3.8 1.5 0.3]], Prediction of  y: [0.]
The probability of sample data point in class 0 is 1.0
The probability of sample data point in class 1 is 0.0
The probability of sample data point in class 2 is 0.0


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  


In [221]:
# iii. Report on the score for kNN, what does the score measure?
score = model.score(X_test, y_test)
print("Score of the model: ", score)

Score of the model:  0.8666666666666667


The score is the coefficient of prediction, it represents the model's capability in predicting the correct label. The best possible score is 1.0, a model with a score closer to 1.0 has high chance to predict correct label.
Our model has score 0.8666666666666667, very close to 1.0, that is suggesting that our model is fitted pretty well to the existing data, our predicted results will show relatively high accuracy.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?